chatkit-core/src/main/kotlin/com/pusher/chatkit/UserSubscription.kt

```diff
@@ -43,195 +33,131 @@ class UserSubscription(
-        chatManager.roomStore -= roomId
-        broadcast(RoomDeleted(roomId))
-    }
-
-    private fun handleRoomUpdated(room: Room) {
-        chatManager.roomStore += room
-        broadcast(RoomUpdated(room))
-    }
-
-    private fun handleRemovedFromRoom(roomId: Int) {
-        chatManager.roomStore -= roomId
-        broadcast(CurrentUserRemovedFromRoom(roomId))
-    }
+    )
 
-    private fun handleAddedToRoom(room: Room) {
-        chatManager.roomStore += room
-        broadcast(CurrentUserAddedToRoom(room))
+    override fun unsubscribe() {
+        subscription.unsubscribe()
+        currentUser?.roomSubscriptions?.forEach(RoomSubscription::unsubscribe)
     }
 
-    private var currentUser: CurrentUser? = null
-
-    private fun handleInitialState(initialState: InitialState) = launch {
-        logger.verbose("Initial state received $initialState")
-        getCursors().mapResult { cursors ->
-            with(currentUser) {
-                when (this) {
-                    null -> CurrentUser(
-                        apiInstance = apiInstance,
-                        createdAt = initialState.currentUser.createdAt,
-                        cursors = cursors.toMutableMap(),
-                        cursorsInstance = cursorsInstance,
-                        id = initialState.currentUser.id,
-                        logger = logger,
-                        filesInstance = filesInstance,
-                        presenceInstance = presenceInstance,
-                        tokenParams = tokenParams,
-                        tokenProvider = tokenProvider,
-                        userStore = userStore,
-                        avatarURL = initialState.currentUser.avatarURL,
-                        customData = initialState.currentUser.customData,
-                        name = initialState.currentUser.name,
-                        updatedAt = initialState.currentUser.updatedAt,
-                        chatManager = chatManager
-                    )
-                    else -> this.also {
-                        it.presenceSubscription.unsubscribe()
-                        it.updateWithPropertiesOf(initialState.currentUser)
+    private fun getCursors(): Future<Result<Map<Int, Cursor>, Error>> =
+        cursorsRequest.mapResult { cursors -> cursors.map { it.roomId to it }.toMap() }
+
+    private fun UserSubscriptionEvent.applySideEffects(): UserSubscriptionEvent = this.apply {
+            when (this) {
+                is InitialState -> {
+                    updateExistingRooms(rooms).forEach(consumeEvent)
+                    chatManager.roomStore += rooms
+                    this@UserSubscription.currentUser?.apply {
+                        close()
+                        updateWithPropertiesOf(currentUser)
                     }
                 }
+                is AddedToRoomEvent -> chatManager.roomStore += room
+                is RoomUpdatedEvent -> chatManager.roomStore += room
```

Comment:
This reads as though the room is being added to the room store - is that the case? Or is it updating the existing room in the room store with the properties of the new "updated" room?

Reply:
Yeah, adding a room to `RoomStore` overrides the current value as it is backed by a `Map<Int, Room>` so it will always override the current value. I can alias a new method to make this more clear.